docs/user-guide/invariants.rst

```diff
@@ -0,0 +1,183 @@
+
+* To run the spec on a correct implementation of :solidity:`ITeams` use
+  :clink:`correct.conf</CVLByExample/Teams/correct.conf>`, which will use the
+  implementation in :clink:`Teams.sol</CVLByExample/Teams/Teams.sol>`.
+* To see the spec discover bugs, use
+  :clink:`buggy.conf</CVLByExample/Teams/buggy.conf>`. This will run the spec
+  against the buggy implementation in
+  :clink:`TeamsBugs.sol</CVLByExample/Teams/TeamsBugs.sol>`.
+
+
+Simple invariants
+-----------------
+
+.. _no_team_for_address_zero:
+
+No team for address zero
+^^^^^^^^^^^^^^^^^^^^^^^^
+We can readily deduce from the properties that :solidity:`teamOf(address(0))` must be
+zero. Here it is written as an invariant:
+
+.. cvlinclude:: ../../Examples/CVLByExample/Teams/Teams.spec
+   :cvlobject: methods addressZeroNotPlayer
+   :caption: :clink:`The team of zero is zero</CVLByExample/Teams/Teams.spec>`
+
+We declared the functions :solidity:`teamOf` and :solidity:`leaderOf` as :cvl:`envfree`
+to remove the need for an :cvl:`env` type argument.
+
+The leader is part of the team
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Another invariant property is that the team-id of the leader of team :math:`x` is
+:math:`x`. This only holds if :math:`x` is not zero and the leader is not
+:solidity:`address(0)`. Here is the property written as an invariant:
+
+.. cvlinclude:: ../../Examples/CVLByExample/Teams/Teams.spec
+   :cvlobject: leaderBelongsToTeam
+   :caption: :clink:`The team's leader is part of the team</CVLByExample/Teams/Teams.spec>`
+
+
+.. index::
+   single: preserved block
+   single: invariant; preserved block
+
+Using preserved blocks inside invariants
+----------------------------------------
+Sometimes additional conditions are needed to prove invariants. These additional
+conditions are given using preserved blocks, see :ref:`preserved`. Here are two
+examples using preserved blocks.
+
+A team not created has no players
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Before team :solidity:`i` is created, :solidity:`leaderOf(i)` must be
+:solidity:`address(0)`. In such a case, there should be no players in team :solidity:`i`.
+We can write this condition as:
+
+.. cvlinclude:: ../../Examples/CVLByExample/Teams/NoPreserved.spec
+   :cvlobject: nonExistTeamHasNoPlayers
+   :caption: :clink:`nonExistTeamHasNoPlayers without preserved block</CVLByExample/Teams/NoPreserved.spec>`
+
+Running this rule, the Prover will find the following violation. The function
+called is :solidity:`changeLeader(address(0))`, changing the leader from address
+:solidity:`a` (where :solidity:`a` is not zero) to zero.
+Before the call :solidity:`address(0)` is a member of team :solidity:`i`, where
+:solidity:`i > 0`. After the call the left hand side of the invariant condition
+holds true: :cvl:`i != 0 && leaderOf(i) == 0`. But the right hand side
+is false for :cvl:`player = a`, since :cvl:`teamOf(a) = i`. The violation is expressed
+in the following table, showing the change in state.
```

Comment:
I think it would also be beneficial to link to a rule report that contains a call-trace of this violation. We need to educate users to read call traces and understand counter examples.

Reply:
Done.